---
title: "TLR expression analysis in *Pleurodeles waltl*"
author: "Krystian Budzik"

output:
  html_document:
    toc: true
    toc_depth: 3
    theme: flatly
    highlight: tango
    css: styles.css
  pdf_document:
    toc: true
    toc_depth: 3
    keep_tex: true
---
# Loading

## Libraries
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(tidyverse)
library(readr)
library(dplyr)
library(tidyr)
library(stringr)
library(gtools)
library(ggplot2)
```
## Files

It is advised to download this directory -  **02_expression** to your local machine to the workspace directory of your RStudio project.

```{r data-loading}
# üìÅ FeatureCounts output
tlr_counts_fc_data <- read_tsv("02_expression/tlr_counts_fc.txt")
tlr_counts_fc_summary <- read_tsv("02_expression/tlr_counts_fc.txt.summary")

# üìÅ BLAST-based annotation table (gene_id ‚Üî TLR name)
anot5_bitscore <- read_csv("03_alignment/anot5_bitscore.csv")

geneid_name <- anot5_bitscore %>%
  select(name, gene_id)

# üìÅ Sample info
Pw_sample_info <- read_tsv("02_expression/Pw_sample_info.txt")
Pw_individual_info <- read_tsv("02_expression/Pw_individual_info.txt")

# ‚úÇÔ∏è Join sample info into one table
Pw_sample_info_2 <- Pw_sample_info %>% left_join(Pw_individual_info, by = "ID")
Pw_sample_info_cut <- Pw_sample_info_2 %>% select(sample, part, stage_descr)
```

# Initial remarks

## GTF file

In this section, the different GTF file was used than in the **Annotation** markdown pipeline. For the sake of the *featureCounts* program, different scaffold naming must be included. This is the only difference, which does not affect compatibility between the work stages.

## Uncertain annotations

Not all the subjects were equally matched with their query (lower than others' bitscore). This is described in the **Annotation** markdown. For the sake of FPKM and TPM computation, this topic is irrelevant. Further discussion of this case can be found in **Annotation** and *Alignment** markdowns.

# featureCounts

For the analysis, the *subread-2.0.6-Linux-x86_64* package was used on the remote server.

## Input

For the input, the **GCF_031143425.1_aPleWal1.hap1.20221129_genomic_scafolds_renamed.gtf** GTF file was searched for *CDS* or *exon* sequences. Subsequently, the following set has been confined only to the records, which include Gene ID from the *anot4_1* file.

```bash
cut -d' ' -f1 anot4_1.txt > gene_id_list.txt

grep -w -E  "CDS|exon" GCF_031143425.1_aPleWal1.hap1.20221129_genomic_scafolds_renamed.gtf > CDS_exon_anot.gtf

grep -Ff gene_id_list.txt CDS_exon_anot.gtf > anot3_2.txt

cp anot3_1.txt prot_id_CDS.gtf
cp anot3_2.txt gene_id_CDS_exon.gtf

echo "Done!" 
```

### CDS and exon records

The GTF file was searched for both CDS and exon records. The *featureCounts* program can be run on both sequence types, with a relevant output in this type of work.

## featureCounts execution

```bash
#!/bin/bash

~/bin/featureCounts \
  -a gene_id_CDS_exon.gtf \
  -o tlr_counts_fc.txt \
  -t CDS \
  -g gene_id \
  -s 0 \
  -p \
  -B \
  --countReadPairs \
  -T 4 \
 ./bams/*.bam

echo "Done!"
```

Where:
```
-a = input file
-o = output file
-t = sequence type (CDS selected)
-g = atribute from the GTF file
-s = strand specificity (0 = both strands count)
-p = paired-end count
-B = only properly mapped pairs
--countReadPairs = pairs counted
-T = CPUt threads
./* = input BAM files
```
### Why CDS, not exon?

CDS attribute was selected to perform the *featureCounts*, instead of the exon. This decision was arbitrary.

CDS coming from the annotation naturally restricts us to the coding sequence of a gene, which is more accurate than the first and last exon boundaries, which include the UTR, which are often poorly annotated.

## Output

*featureCounts* execution resulted in two files: 1. **tlr_counts_fc.txt** and 2. **tlr_counts_fc.txt.summary**. The first one is the

Read number aligned with a gene.

The second one is a summary, where the information about the number of paired, unpaired, unmapped, etc., reads can be found. Both ones are going to be used in subsequent steps of the expression analysis.


# Output files processing

Both *featureCounts* files were downloaded from the server and loaded into RStudio

```{r, eval=FALSE}
tlr_counts_fc_data <- read_tsv("tlr_counts_fc.txt")
tlr_counts_fc_summary <- read_tsv("tlr_counts_fc.txt.summary")
```

## fc_data

### Cleaning from the irrelevant naming

```{r, eval=FALSE}
library(dplyr)

tlr_counts_fc_data <- rename(tlr_counts_fc_data, gene_id = Geneid)
tlr_counts_fc_data$gene_id <- gsub("LOC", "", tlr_counts_fc_data$gene_id)
```

### Adding gene ID instead of gene name

Some records were automatically annotated as TLR2, TLR3, etc. For clarity purposes, these gene names were replaced by the proper gene ID of each gene.

```{r, eval=FALSE}
library(dplyr)
tlr_counts_fc_data$gene_id <- gsub("TLR2", "138303579", tlr_counts_fc_data$gene_id)
tlr_counts_fc_data$gene_id <- gsub("TLR3", "138245949", tlr_counts_fc_data$gene_id)
tlr_counts_fc_data$gene_id <- gsub("TLR5", "138295498", tlr_counts_fc_data$gene_id)
tlr_counts_fc_data$gene_id <- gsub("TLR8", "138250171", tlr_counts_fc_data$gene_id)
```

### Adding the gene name column

During the work, the uncertain annotation of the TLR01 gene described in the **Alignment** markdown was identified as the TLR06. This change was applied to the annotation dataset used to join the gene name to the rest of the expression data.

```{r, eval=FALSE}

library(dplyr)

geneid_name <- anot5 %>% select(1, 3) %>%
  mutate(name = if_else(gene_id == "138247579", "TLR06", name))

tlr_counts_fc_names <- tlr_counts_fc_data %>%
  left_join(geneid_name, by = "gene_id")
```

## fc_summary

### Summing the features

For the subsequent computational steps, a sum of the features in each sample was needed. The sum was defined as *Assigned* + *unassigned_noFeatures*, as these two categories include the relevant reads only, omitting unmapped and unpaired ones.

```{r, eval=FALSE}

features_sum <- tlr_counts_fc_summary %>%
  filter(Status %in% c("Assigned", "Unassigned_NoFeatures"))

numeric_cols <- features_sum[, -1] %>% mutate_all(as.numeric)

total_row <- colSums(numeric_cols)

total_named <- c(Status = "Features_sum", total_row)

features_sum <- bind_rows(
  features_sum,
  as.data.frame(t(total_named), stringsAsFactors = FALSE)
)
```

## Joining in the long format

The master table for the FPKM and TPM computation was prepared by formatting the datasets to the long format and joining them together. 

```{r, eval=FALSE}
library(dplyr)
library(tidyr)

tlr_long <- tlr_counts_fc_names %>%
  pivot_longer(
    cols = matches("\\.bam$"), 
    names_to = "sample",
    values_to = "counts"
  )

summary_long <- features_sum %>% slice(c(3)) %>%
  pivot_longer(
    cols = matches("\\.bam$"),
    names_to = "sample",
    values_to = "features_sum"
  ) %>% select(sample, features_sum)

tlr_long_total <- tlr_long %>% left_join(summary_long, by = "sample")
```

# FPKM and TPM

## FPKM computation

```{r, eval=FALSE}
library(dplyr)

tlr_fpkm_all <- tlr_long_total %>%
  mutate(
    length_kb = Length / 1000,
    rpk = counts / length_kb,
    fpkm = (rpk / features_sum) * 1e6
  ) 

tlr_fpkm <- tlr_fpkm_all %>% select(name, gene_id, sample, fpkm)

```

### Preparing the FPKM set to join the full data set
```{r, eval=FALSE}
tlr_fpkm_join <- tlr_fpkm %>% mutate(sample = str_extract(sample, "Pw[^.]+")) %>%
  select(gene_id, sample, fpkm)
```

## TPM computation

## TPM

```{r, eval=FALSE}
tlr_rpk_sum <- tlr_fpkm_all %>%
  group_by(sample) %>%
  summarise(total_rpk = sum(rpk, na.rm = TRUE))

tlr_long_total <- tlr_long_total %>%
  left_join(tlr_rpk_sum, by = "sample")
```


```{r, eval=FALSE}

tlr_tpm_all <- tlr_long_total %>%
  mutate(
    length_kb = Length / 1000,
    rpk = counts / length_kb,
    scalling_factor = total_rpk.x / 1e6,
    tpm = (rpk / scalling_factor),
    tpm_percent = (tpm / 1e4)
  ) 

tlr_tpm_all <- tlr_tpm_all %>% select(-c(total_rpk.x, total_rpk.y))

tlr_tpm <- tlr_tpm_all %>% select(name, gene_id, length_kb, sample, tpm, tpm_percent)
```

# Sample info

### Loading

```{r, eval=FALSE}
library(tidyverse)
Pw_sample_info <- read_tsv("Pw_sample_info.txt")
Pw_individual_info <- read_tsv("Pw_individual_info.txt")
Pw_sample_info_2 <- Pw_sample_info %>% left_join(Pw_individual_info, by = "ID")
Pw_sample_info_cut <- Pw_sample_info_2 %>% select(sample, part, stage_descr)
```

### Joining
```{r, eval=FALSE}
tlr_tpm_description <- tlr_tpm %>% mutate(sample = str_extract(sample, "Pw[^.]+")) %>%  
  
  left_join(Pw_sample_info_cut, by = "sample")

```

### Master table preparation

```{r, eval=FALSE}
library(dplyr)

tlr_tpm_fpkm_dataset <- tlr_tpm_description %>%
  left_join(tlr_fpkm_join, by = c("sample", "gene_id")) %>%

  # Subgroups ‚Äì ontogenetic stages used in the visualisation
  mutate(dev_subgroup = case_when(
    stage_descr %in% c("three_fingers", "three_toes") ~ "Young larva",
    stage_descr %in% c("four_toes", "five_toes", "full_dev") ~ "Older larva",
    stage_descr %in% c("pre_meta", "mid_meta", "post_meta") ~ "Metamorphosis",
    stage_descr == "adult" ~ "Adult",
    TRUE ~ stage_descr
  )) %>%

  # General groups - not used
  mutate(dev_group = case_when(
    dev_subgroup %in% c("Young larva", "Older larva") ~ "Larva",
    dev_subgroup == "Metamorphosis" ~ "Metamorphosis",
    dev_subgroup == "Adult" ~ "Adult",
    TRUE ~ "other"
  )) %>%

  # Setting the right order of the categories
  mutate(
    dev_subgroup = factor(dev_subgroup, levels = c("Young larva", "Older larva", "Metamorphosis", "Adult")),
    dev_group = factor(dev_group, levels = c("Larva", "Metamorphosis", "Adult"))
  )
  
```

### Preparing a short name category for TLR

The short name category is going to be used in the data visualization using the *ggplot2* library.

```{r, eval=FALSE}
library(dplyr)

tlr_tpm_fpkm_dataset <- tlr_tpm_fpkm_dataset %>%
  mutate(tlr_short0 = gsub("^TLR", "", name)) # Removes "TLR"
  
  tlr_tpm_fpkm_dataset <- tlr_tpm_fpkm_dataset %>%
  mutate(tlr_short = gsub("0", "", tlr_short0)) %>% # Removes "0"
    select(-c("tlr_short0"))
```

### Setting the right order of the TLR names

```{r, eval=FALSE}
library(gtools)

# Sorting
tlr_order <- mixedsort(unique(tlr_tpm_fpkm_dataset$tlr_short))

# Factor
tlr_tpm_fpkm_dataset <- tlr_tpm_fpkm_dataset %>%
  mutate(tlr_short = factor(tlr_short, levels = tlr_order))

# Check
levels(tlr_tpm_fpkm_dataset$tlr_short)
```

# Visualisation

## FPKM-based global TLR expression plot

### FPKM sum and mean FPKM in each developmental group

```{r, eval=FALSE}
library(dplyr)
library(ggplot2)

# FPKM sum for each sample
sample_fpkm_summary <- tlr_tpm_fpkm_dataset %>%
  group_by(sample, dev_subgroup) %>%
  summarise(sample_sum_fpkm = sum(fpkm, na.rm = TRUE), .groups = "drop")

# Mean FPKM in each developmental group
subgroup_summary <- sample_fpkm_summary %>%
  group_by(dev_subgroup) %>%
  summarise(
    mean_fpkm = mean(sample_sum_fpkm),
    sd_fpkm = sd(sample_sum_fpkm),
    n = n()
  )
```

### Global expression plot
```{r, eval=FALSE}

# Colours
group_colors <- c(
  "Young larva" = "#66c2a5",
  "Older larva" = "#fc8d62",
  "Metamorphosis" = "#8da0cb",
  "Adult" = "#ce4e9b"
)

# Plot
library(ggplot2)

ggplot(sample_fpkm_summary, aes(x = dev_subgroup, y = sample_sum_fpkm, color = dev_subgroup)) +
  geom_boxplot(outlier.shape = NA, fill = NA, size = 1) + 
  geom_jitter(width = 0.2, size = 3, alpha = 0.8) + 
  stat_summary(fun = mean, geom = "point", color = "black", shape = 18, size = 4) +
  labs(
    title = NULL,
    x = NULL,
    y = "Total FPKM",
    color = "Stage"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(size = 13),
    axis.title.x = element_text(size = 15, margin = margin(t = 15)),
    axis.title.y = element_text(size = 15, margin = margin(r = 15)),
    legend.position = "none"
    
  )
```

## TPM-based relative expression plot

```{r}
library(ggplot2)

tlr_order <- c("1", "2", "3", "5", "5L", "6", "7", "8", "9A", "9B", "13", "14", "19", "21", "22A")

tlr_tpm_fpkm_dataset <- tlr_tpm_fpkm_dataset %>%
  mutate(tlr_short = factor(tlr_short, levels = tlr_order))

tlr_colors["6"] <- "#a6cee3"  # Soft blue, nice for differentiation


tlr_tpm_fpkm_dataset$dev_subgroup <- factor(
  tlr_tpm_fpkm_dataset$dev_subgroup,
  levels = c("Young larva", "Older larva", "Metamorphosis", "Adult"))

p2_final <- ggplot(tlr_tpm_fpkm_dataset,
                   aes(x     = tlr_short,
                       y     = tpm_percent,
                       fill  = tlr_short,
                       color = tlr_short)) +
  
  # boxplot, jitter
  geom_boxplot(width = 0.6, outlier.size = 0, alpha = 0.1, size = 0.6) +
  geom_jitter(shape = 21, size = 2, width = 0, alpha = 0.7, show.legend = FALSE) +
  
  # panels
  facet_wrap(~ dev_subgroup,
             ncol           = 1,
             strip.position = "right",
             drop           = FALSE) +

  # scale and colours
  scale_x_discrete(limits = levels(tlr_tpm_fpkm_dataset$tlr_short)) +
  scale_fill_manual(values = tlr_colors) +
  scale_color_manual(values = tlr_colors) +
  scale_y_continuous(trans  = "sqrt",
                     breaks = c(0, 2, 10, 20, 50),
                     expand = expansion(mult = c(0, 0.05))) +

  # descriptions
  labs(x = "TLR gene",
       y = "TPM (%)") +


  theme_minimal(base_size = 14) +
  theme(
    # stripy po prawej stronie w stylu jak wcze≈õniej
    strip.placement        = "outside",
    strip.background       = element_rect(fill = NULL, color = NA),
    strip.text.y.right     = element_text(
                               angle  = 270,
                               size   = 14,
                               face   = NULL,
                               color  = "black",
                               vjust  = 0.5,
                               margin = margin(l = 10)  # odstƒôp od wykresu
                             ),

    # Vertical
    panel.grid.major.y     = element_line(color = "gray90", size = 0.3),
    panel.grid.major.x     = element_line(color = "gray90", size = 0.3),
    panel.grid.minor       = element_blank(),

    # odstƒôpy i proporcje
    panel.spacing.y        = unit(6, "mm"),
    panel.spacing.x        = unit(5, "mm"),
    

    # margins
    plot.margin            = unit(rep(4, 4), "mm"),

    # axis
    axis.text.x            = element_text(angle = 0, hjust = 0.5, size = 13),
    axis.text.y            = element_text(size = 12),
    axis.title.x           = element_text(size = 14, margin = margin(t = 10)),
    axis.title.y           = element_text(size = 14, margin = margin(r = 10)),

    # No legend
    legend.position        = "none"
  )


```
## fish-type TLRs TPM-based plot

```{r, eval=TRUE}

# Assuming you already have `tlr_tpm_fpkm_dataset` loaded
# Filter only for the fish-specific TLRs

fish_tlrs <- c("14", "19", "21", "22A")

tlr_fish_data <- tlr_tpm_fpkm_dataset %>%
  filter(tlr_short %in% fish_tlrs) %>%
  mutate(
    dev_subgroup = factor(dev_subgroup, levels = c("Young larva", "Older larva", "Metamorphosis", "Adult")),
    tlr_label = factor(paste0("TLR", tlr_short), levels = paste0("TLR", fish_tlrs))
  )


# Plot
p_fish <- ggplot(tlr_fish_data, aes(x = dev_subgroup, y = tpm_percent)) +
  geom_jitter(aes(color = part), size = 2.5, width = 0.2, alpha = 0.7) +
  stat_summary(aes(group = 1), fun = mean, geom = "line", color = "black", size = 0.7) +
  stat_summary(aes(group = 1), fun = mean, geom = "point", shape = 21, fill = "white", color = "black", size = 3.5) +
  scale_color_manual(values = c(
  "lungs" = "#1b9e77",
  "tail" = "#d95f02",
  "spleen" = "#7570b3",
  "intestine" = "#e7298a",
  "liver" = "#66a61e",
  "heart" = "#e6ab02",
  "head" = "#a6761d",
  "middle" = "#666666",
  "organs" = "#1f78b4",
  "middle_skin_spine" = "#b2182b",
  "middle_organs" = "#542788"
)) +

  facet_wrap(~ tlr_label, ncol = 2) +
  labs(
    title = NULL,
    x = NULL,
    y = "Expression (TPM %)",
    color = "Organ"
  ) +
  theme_minimal(base_size = 15) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.title.y = element_text(size = 14, margin = margin(r = 12)),
    panel.grid.minor.y = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(size = 13),
    legend.text = element_text(size = 12),
    strip.text = element_text(size = 14, face = "bold")
  )


```
